### PR TITLE
Heroku Integration

### PR DESCRIPTION
**Pull request by awan7676**

### CODE REVIEW

File: apps/heroku/src/app/oauth/service.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        data: {
+          isFirstSync: true,
+          organisationId,
+          region,
+          syncStartedAt: Date.now(),
+          range: null,
+        },
+      },
+    ]);
+    return organisation;
```
Comment (elbalexandre) [inline apps/heroku/src/app/oauth/service.ts:61]:
```suggestion
```
We are not using it in `route.ts` so we can remove it.
If it's use for testing purpose, we can assume in it that if the function does not throws (resolves.toBeUndefined()), it's a success.

File: apps/heroku/src/connectors/auth.test.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    );
+  });
+
+  test('should not throw when authorization code is valid', async () => {
+    try {
+      await getAccessToken(validAuthCode);
+      expect(true).toBe(true);
+    } catch (error) {
+      expect(error).toBeNull();
+    }
```
Comment (elbalexandre) [inline apps/heroku/src/connectors/auth.test.ts:38]:
```suggestion
      await expect(getAccessToken(validAuthCode)).resolves.toStrictEqual({
        accessToken: '...',
        refreshToken: '...',
        expiresIn: '...'
      });
```

Comment (elbalexandre) [reply apps/heroku/src/connectors/auth.test.ts:38]: Could you apply the same comments on every test case: use `.resolves` or `.rejects.toBeInstanceOf(HerokuError)`

File: apps/heroku/src/inngest/functions/users/delete-user.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    retries: env.REMOVE_ORGANISATION_MAX_RETRY,
+  },
+  {
+    event: 'heroku/users.delete.requested',
+  },
+  async ({ event, step }) => {
+    const { id, organisationId } = event.data;
+
+    // retrieve the Heroku organisation access token and team Id
+    const [accessToken, teamId] = await step.run('get-access-token', async () => {
```
Comment (elbalexandre) [inline apps/heroku/src/inngest/functions/users/delete-user.ts:24]: Let's make the step title clearer
```suggestion
    const organisation = await step.run('get-organisation', async () => {
```

File: apps/heroku/src/inngest/functions/users/delete-user.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        .select({
+          accessToken: Organisation.accessToken,
+          teamId: Organisation.teamId,
+        })
+        .from(Organisation)
+        .where(eq(Organisation.id, organisationId));
+      if (!organisation) {
+        throw new NonRetriableError(`Could not retrieve organisation with id=${organisationId}`);
+      }
+      return [organisation.accessToken, organisation.teamId];
```
Comment (elbalexandre) [inline apps/heroku/src/inngest/functions/users/delete-user.ts:35]:
```suggestion
      return organisation
```

File: apps/heroku/src/inngest/functions/users/delete-user.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    });
+
+    if (accessToken && teamId) {
+      await step.run('delete-user', async () => {
+        await deleteUser(accessToken, teamId, id);
+      });
+    }
```
Comment (elbalexandre) [inline apps/heroku/src/inngest/functions/users/delete-user.ts:42]: We don't need the if statement here
```suggestion
      await step.run('delete-user', async () => {
        await deleteUser(organisation.accessToken, organisation.teamId, id);
      });
```

File: apps/heroku/src/inngest/functions/users/delete-user.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+    return {
+      status: 'completed',
+    };
```
Comment (elbalexandre) [inline apps/heroku/src/inngest/functions/users/delete-user.ts:46]:
```suggestion
```

If it throw it fails, if not it succeed: so we don't need a status here.

File: apps/heroku/src/inngest/functions/users/sync-users-page.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+    const elba = new Elba({
+      organisationId,
+      apiKey: env.ELBA_API_KEY,
+      baseUrl: env.ELBA_API_BASE_URL,
+      region,
+    });
+
+    // retrieve the Heroku API Access Token and team Id
+    const [accessToken, teamId] = await step.run('get-access-token', async () => {
```
Comment (elbalexandre) [inline apps/heroku/src/inngest/functions/users/sync-users-page.ts:51]: Same comment as the other inngest function + organisation.accessToken & teamId should not be nullable so no need to check them in `list-users`.

Comment (awan7676) [reply apps/heroku/src/inngest/functions/users/sync-users-page.ts]: Done

File: apps/heroku/src/inngest/functions/users/delete-user.ts (hunk 6 above)
Comment (awan7676) [reply apps/heroku/src/inngest/functions/users/delete-user.ts]: Done

File: apps/heroku/src/inngest/functions/users/delete-user.ts (hunk 5 above)
Comment (awan7676) [reply apps/heroku/src/inngest/functions/users/delete-user.ts]: Done

File: apps/heroku/src/inngest/functions/users/delete-user.test.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  refreshToken: 'refresh-token',
+  teamId: 'team-id',
+  region: 'us',
+};
+
+const userId = 'user-id';
+
+const setup = createInngestFunctionMock(deleteHerokuUser, 'heroku/users.delete.requested');
+
+describe('delete-user-request', () => {
```
[the remark this answers is not in the document]
Comment (awan7676) [reply apps/heroku/src/inngest/functions/users/delete-user.test.ts:21]: Done

File: apps/heroku/src/connectors/teams.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+    throw new HerokuError('Failed to fetch', { response });
+  }
+  const data = (await response.json()) as HerokuTeam[];
+  let teamId: string | null | undefined = null;
+
+  //extract first team of the enterprise account
+  if (data.length > 0) {
+    teamId = data[0]?.id;
+  }
+  return teamId;
```
[the remark this answers is not in the document]
Comment (awan7676) [reply apps/heroku/src/connectors/teams.ts]: Done

File: apps/heroku/src/connectors/auth.test.ts (hunk 2 above)
Comment (awan7676) [reply apps/heroku/src/connectors/auth.test.ts]: Done

File: apps/heroku/src/app/oauth/service.ts (hunk 1 above)
Comment (awan7676) [reply apps/heroku/src/app/oauth/service.ts]: Done

File: apps/heroku/src/app/oauth/service.ts [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+};
+
+export const setupOrganisation = async ({
+  organisationId,
+  code,
+  region,
+}: SetupOrganisationParams) => {
+  const { accessToken, refreshToken, expiresIn } = await getAccessToken(code);
+  const teamId = await getTeamId(accessToken);
+  if (teamId) {
```
[the remark this answers is not in the document]
Comment (awan7676) [reply apps/heroku/src/app/oauth/service.ts]: Done

**Comment by elbalexandre** [comment]
I am closing this PR. I had to make changes to make it works and it was very far from staging, it has required a lot of config changes, so I directly push to the target branch.

**Pull request closed without merging**